### ISSUE TITLE
Doubt on sampling methods

### ISSUE THREAD

**Issue by ZcharlieZ** [joshspeagle/dynesty#472]
Hi all!

I do have a doubt about how to plot the results after sampling. Let's suppose I would like to fit a linear relation. After the sampling, I have the posterior of all the parameters of the model. This results in having N curves. Thus at a fixed x value, I can compute for instance the median and the 16% and 84% around the curves. In the case of dynesty, however, weights are used. The posterior should be weighted and, if not wrong, the weight is logwt-logz[-1] for the PDFs in dyplot traceplots or cornerplots.
If everything above is correct, how can I plot the median and the uncertainties of the model?

Many thanks in advance

**Comment by joshspeagle** [comment]
You can also use the `quantile` function in `utils` to do exactly this, which computes quantiles following a weighted CDF using the ordered samples.

**Comment by ZcharlieZ** [comment]
Hi all!

Thanks for your replies. Let me go a bit more into the details of what I'm doing. I'm sampling the PDFs using the Dynamic Nested Sampler. The configuration is the following:

```
dsampler = dynesty.DynamicNestedSampler(log_likelihood, prior_transform, ndim=len(parameter_labels),
                                    bound='multi', sample='rslice', rstate=rstate, pool=pool, queue_size=250, nlive=1000)
dsampler.run_nested(dlogz_init=0.01, print_progress=True)
```

The first thing I noticed is that, keeping everything the same and using the same random state (rstate=np.random.default_rng(18)), but changing the sample method from _rslice_ to _rwalk_ the PDFs change significantly. Albeit the number of parameters used is lower than 10, for which _unif_ is suggested, I'm using one of the two methods above because sampling with _unif_ is very slow. Here below you find the traceplots using _rslice_ and _rwalk_, respectively. In both cases, I also noticed that, after starting the run, the dlogz value in output from print_progress has a starting value, then for a while it becomes a way larger, and finally it reduces to reach the convergence below the threshold (dlogz_init=0.01).

**rslice**
![image](https://github.com/joshspeagle/dynesty/assets/33525059/fee809db-26d6-4210-adfa-03be439da424)

**rwalk**
![da517e7a-a469-4110-9562-e5c14de1699c](https://github.com/joshspeagle/dynesty/assets/33525059/3765af75-9862-41d9-ab8b-145e2f192833)

The second sampling should be more in line with what I should expect. In addition, I tried to sample the PDFs using an MCMC algorithm. In the latter, the PDFs differ from the "rwalk" case, but they look a bit more Gaussian-like distributed.

Finally, coming to the use of weights, the PDFs of the parameters differ considerably wider when I get the weighted samples. Indeed, when I plot the final relation, the 1sigma uncertainty around the curve is very large. Here below I attach the corner plots and the relations for the unweighted (blue colours) and weighted (red colours) samples.

![e67972a5-da9e-4eb5-aeb1-b8644025d11b](https://github.com/joshspeagle/dynesty/assets/33525059/867c75e2-84c1-48bb-a3fc-e2634e33d389)

![5d6a7450-33d1-4fa2-a367-583adb1dd723](https://github.com/joshspeagle/dynesty/assets/33525059/ad6a047f-83e6-4339-aa1d-f4e00bedb4cb)

To summarise, I list my questions in the following.

1. What is the reason why using a different sample method dramatically changes the results? Given that, which is the method one should use? 
2. Using an MCMC algorithm, similar to using a different sample method in Dynesty, leads to different results and perhaps more regular (looking both at the chains and corner plots). What might be the reason?
3. If I understand correctly the scope of weights, I would expect the use of weighted samples to give a more "accurate" estimate of the posterior thus reducing the uncertainty around my relation, but this is not my case. Maybe I'm wrong about it. Could you explain me better this aspect, please?

Thank you very much for your help, very appreciated!

_Carlo_

**Comment by segasai** [comment]
Hi, 

* The different sampling methods could give you somewhat different results if the MH part of sampling is not efficient and/or number of steps in the random walk is too low. If you are seeing that you probably want to increase the number of points and/or increase the length of the random walk steps. In your case, from what I can see you get completely different values and it seems your posterior is strongly multimodal, in that case, you need to ensure to have enough live points to ensure you discover all the modes. 
* Regarding weights. I don't quite understand what you are trying to say. The are only two *correct* ways of using posteriors from dynesty. One using all samples with weights. Another one is use equally weighted samples (from samples_equal). Just taking original samples and ignoring weights is not meaningful.

**Comment by ZcharlieZ** [comment]
Hi @segasai,

thanks for your previous reply. I reply to your previous points here below.

- For what concerns the weighted sample, I was expecting the PDFs of parameters to be tighter than those of the non-weighted sample because in the latter all the live points are equally considered, even those far from the max likelihood region. Instead, it is totally the opposite.

- Following your suggestion about increasing the number of live points, I performed some tests. In each test, using NestedSampler, I kept fixed the following parameters: bound='multi', rstate=np.random.default_rng(18), queue_size=256, dlogz=0.01. What I varied were both the sampling method ('rwalk' or 'rslice') and the number of live points (4096, 8192, 16384).
<img width="1716" alt="image" src="https://github.com/joshspeagle/dynesty/assets/33525059/f3fd0a89-f59b-485c-8dbf-4fc449ef4223">
<img width="1716" alt="image" src="https://github.com/joshspeagle/dynesty/assets/33525059/a354456e-6eb9-4431-847e-e3cabfe7769a">
<img width="1716" alt="image" src="https://github.com/joshspeagle/dynesty/assets/33525059/6a900792-ce7a-430d-b2d1-f1aceb5ca150">
<img width="1716" alt="image" src="https://github.com/joshspeagle/dynesty/assets/33525059/6a08cf41-3359-48bf-ba20-0fc38e8c222e">

As clearly visible, all the PDFs are dramatically different from each other, regardless of the number of live points or the sampling method. Do you think I still need to increase even more the number of live points? Apparently, I am not seeing any improvement.

Many thanks in advance for the help.

_Carlo_
